finality-aleph/src/network.rs

```diff
@@ -87,51 +124,95 @@ impl<B: BlockT, H: ExHashT> Network<B> for Arc<NetworkService<B, H>> {
+    authentications: HashMap<SessionId, NodeIndex>,
+}
 
 impl PeerInfo {
     fn new() -> Self {
-        PeerInfo {}
+        PeerInfo {
+            authentications: HashMap::new(),
+        }
+    }
+
+    fn authenticated_for(&self, session_id: &SessionId) -> bool {
+        self.authentications.get(session_id).is_some()
+    }
+
+    fn authenticate(&mut self, session_id: SessionId, node_id: NodeIndex) {
+        self.authentications.insert(session_id, node_id);
+    }
+
+    fn iter(&self) -> impl Iterator<Item = (&SessionId, &NodeIndex)> {
+        self.authentications.iter()
     }
 }
 
-#[derive(Debug)]
 struct Peers {
-    pub(crate) peers: HashMap<PeerId, PeerInfo>,
+    all_peers: HashMap<PeerId, PeerInfo>,
+    to_peer: HashMap<(SessionId, NodeIndex), PeerId>,
 }
 
 impl Peers {
     pub(crate) fn new() -> Self {
         Peers {
-            peers: HashMap::new(),
+            all_peers: HashMap::new(),
+            to_peer: HashMap::new(),
         }
     }
+
     pub(crate) fn insert(&mut self, peer: PeerId) {
-        self.peers.insert(peer, PeerInfo::new());
+        self.all_peers.insert(peer, PeerInfo::new());
     }
 
-    pub(crate) fn remove(&mut self, peer: &PeerId) {
-        self.peers.remove(peer);
+    pub(crate) fn authenticated(&self, peer: &PeerId, session_id: &SessionId) -> bool {
+        match self.all_peers.get(peer) {
+            Some(info) => info.authenticated_for(session_id),
+            None => false,
+        }
     }
 
-    pub(crate) fn _contains(&self, peer: &PeerId) -> bool {
-        self.peers.contains_key(peer)
+    pub(crate) fn authenticate(
+        &mut self,
+        peer: &PeerId,
+        session_id: SessionId,
+        node_id: NodeIndex,
+    ) {
+        if self.all_peers.get(peer).is_none() {
+            self.insert(peer.clone());
+        }
+        self.all_peers
+            .get_mut(peer)
+            .unwrap()
+            .authenticate(session_id, node_id);
+        self.to_peer.insert((session_id, node_id), peer.clone());
     }
 
-    //TODO: optimize this (it does not need to be perfectly random, if this helps)
-    pub(crate) fn sample_random(&self) -> Option<PeerId> {
-        let peers: Vec<&PeerId> = self.peers.keys().collect();
-        let mut rng = thread_rng();
-        peers.choose(&mut rng).cloned().cloned()
+    pub(crate) fn remove(&mut self, peer: &PeerId) {
+        if let Some(peer_info) = self.all_peers.get(peer) {
+            for (session_id, node_id) in peer_info.iter() {
+                self.to_peer.remove(&(*session_id, *node_id));
+            }
+        }
+        self.all_peers.remove(peer);
     }
 
-    pub(crate) fn iter(&self) -> impl Iterator<Item = (&PeerId, &PeerInfo)> {
-        self.peers.iter()
+    pub(crate) fn iter(&self, session_id: SessionId) -> impl Iterator<Item = &PeerId> {
+        self.to_peer
+            .iter()
+            .filter(move |((s_id, _), _)| *s_id == session_id)
+            .map(|(_, p_id)| p_id)
```

Comment:
`iter` is a bit unfortunate name. How about `authenticated_peers_for`?
Maybe we would like to make  `to_peer` a `HashMap<SessionId, HashMap<NodeIndex, PeerId>>` to avoid iterating the whole hashmap?

Reply:
Yup, that's definitely better. Unfortunately some of the code became potentially less readable due to the half-baked nature of functional features (well, maybe 7/8ths-baked, but still).